pkg/storegateway/bucket_chunk_reader.go

```diff
@@ -242,6 +243,132 @@ type loadIdx struct {
+	)
+	if seq >= uint32(len(r.toLoad)) {
+		return errors.Errorf("reference sequence %d out of range", seq)
+	}
+
+	var totalLen int64
+	for _, c := range g.refs {
+		totalLen += int64(c.length)
+	}
+
+	r.toLoad[seq] = append(r.toLoad[seq], rangeLoadIdx{offset: off, seriesEntry: seriesEntry, rangeEntry: rangeEntry, estLen: totalLen})
+	return nil
+}
+
+func (r *bucketChunksRangesReader) loadRanges(partialSeries []partialSeriesChunksSet, stats *safeQueryStats) error {
+	g, ctx := errgroup.WithContext(r.ctx)
+	s, ctx := tracing.StartSpan(ctx, "loadRanges")
+	defer s.Finish()
+
+	for seq, pIdxs := range r.toLoad {
+		sort.Slice(pIdxs, func(i, j int) bool {
+			return pIdxs[i].offset < pIdxs[j].offset
+		})
+		parts := r.block.partitioners.chunks.Partition(len(pIdxs), func(i int) (start, end uint64) {
+			return uint64(pIdxs[i].offset), uint64(pIdxs[i].offset) + uint64(pIdxs[i].estLen)
+		})
+
+		for _, p := range parts {
+			seq := seq
+			p := p
+			indices := pIdxs[p.ElemRng[0]:p.ElemRng[1]]
+			g.Go(func() error {
+				return r.loadChunksRanges(ctx, partialSeries, seq, p, indices, stats)
+			})
+		}
+	}
+	return g.Wait()
+}
+
+// loadChunksRanges will read the part [start, end] from the segment file with sequence number seq.
+// This data range covers ranges starting at supplied chunks ranges.
+func (r *bucketChunksRangesReader) loadChunksRanges(ctx context.Context, partialSeries []partialSeriesChunksSet, seq int, part Part, ranges []rangeLoadIdx, stats *safeQueryStats) error {
+	// Get a reader for the required range.
+	reader, err := r.block.chunkRangeReader(ctx, seq, int64(part.Start), int64(part.End-part.Start))
+	if err != nil {
+		return errors.Wrap(err, "get range reader")
+	}
+	defer runutil.CloseWithLogOnErr(r.block.logger, reader, "readChunkRange close range reader")
+	bufReader := bufio.NewReaderSize(reader, mimir_tsdb.EstimatedMaxChunkSize)
+
+	// Since we may load many chunks, to avoid having to lock very frequently we accumulate
+	// all stats in a local instance and then merge it in the defer.
+	localStats := queryStats{}
+	defer stats.merge(&localStats)
+
+	readOffset := int(ranges[0].offset)
+
+	for rIdx, rng := range ranges {
+		// Fast forward range reader to the next chunk range start in case of sparse byte range.
+		for readOffset < int(rng.offset) {
+			n, err := io.CopyN(io.Discard, bufReader, int64(rng.offset)-int64(readOffset))
+			if err != nil {
+				return fmt.Errorf("fast forward range reader (block %s segment %d offset %d read offset %d): %w", r.block.meta.ULID, seq, rng.offset, readOffset, err)
+			}
+			readOffset += int(n)
+		}
+		rLen := int(rng.estLen)
+		if rIdx+1 < len(ranges) {
+			if diff := ranges[rIdx+1].offset - rng.offset; int(diff) < rLen {
+				// When a range contains a chunk with an estimated length, a range can overlap with the next one.
+				// In this case we want to clamp it to beginning of the next range.
+				rLen = int(diff)
+			}
+		}
+		rangeBuf := make([]byte, rLen)
```

Comment:
[for a follow up discussion] Have you experienced with a memory pool here?

Reply:
i ran some tests with a SlabPool, but i don't think the results were very reliable. The overall theme was that running with 16K slab size was too small and there were about 1/4 of allocations happening because the slab was too small for the range bytes. I also tried with 1.6M and that was too large, i suspect because of too few selected series. 

I haven't tried with any values between 1.6M and 16K, maybe the sweet spot is somewhere there.

Reply:
We can reconsider a pool here later, if we see this popping up in profiles.